include/sparrow/dynamic_bitset.hpp

```diff
@@ -380,4 +545,239 @@ namespace sparrow
+    }
+
+    template <class B, bool is_const>
+    void bitset_iterator<B, is_const>::decrement()
+    {
+        // decreasing moves to previous block
+        if (is_first_bit_of_block(m_index))
+        {
+            --p_block;
+            m_index = sizeof(block_type);
+        }
+        else
+        {
+            --m_index;
+        }
+    }
+
+    template <class B, bool is_const>
+    void bitset_iterator<B, is_const>::advance(difference_type n)
+    {
+        if (n >= 0)
+        {
+            if (m_index + n < bitset_type::s_bits_per_block)
+            {
+                m_index += n;
+            }
+            else
+            {
+                size_type to_next_block = bitset_type::s_bits_per_block - m_index;
+                n -=  to_next_block;
+                size_type block_n = n / bitset_type::s_bits_per_block;
+                p_block += block_n + 1;
+                n -= block_n * bitset_type::s_bits_per_block;
+                m_index = n;
+            }
+        }
+        else
+        {
+            difference_type mn = -n;
+            if (m_index >= mn)
+            {
+                m_index -= mn;
+            }
+            else
+            {
+                size_type block_n = mn / bitset_type::s_bits_per_block;
+                p_block -= block_n;
+                mn -= block_n * bitset_type::s_bits_per_block;
+                if (m_index >= mn)
+                {
+                    m_index -= mn;
+                }
+                else
+                {
+                    mn -= m_index;
+                    --p_block;
+                    m_index = bitset_type::s_bits_per_block - mn;
+                }
+            }
+        }
+    }
+
+    template <class B, bool is_const>
+    auto bitset_iterator<B, is_const>::distance_to(const self_type& rhs) const -> difference_type
+    {
+        if (p_block == rhs.p_block)
+        {
+            return rhs.m_index - m_index;
+        }
+        else
+        {
+            auto dist1 = distance_to_begin();
+            auto dist2 = rhs.distance_to_begin();
+            return dist2 - dist1;
+        }
+    }
+
+    template <class B, bool is_const>
+    bool bitset_iterator<B, is_const>::equal(const self_type& rhs) const
+    {
+        return p_block == rhs.p_block && m_index == rhs.m_index;
+    }
+
+    template <class B, bool is_const>
+    bool bitset_iterator<B, is_const>::less_than(const self_type& rhs) const
+    {
+        return (p_block < rhs.p_block) || (p_block == rhs.p_block && m_index < rhs.m_index);
+    }
+
+    template <class B, bool is_const>
+    bool bitset_iterator<B, is_const>::is_first_bit_of_block(size_type index) const
+    {
+        return m_index % bitset_type::s_bits_per_block == 0;
+    }
+
+    template <class B, bool is_const>
+    auto bitset_iterator<B, is_const>::distance_to_begin() const -> difference_type
+    {
+        return bitset_type::s_bits_per_block * (p_block - p_bitset->begin().p_block) + m_index;
```

Comment:
I believe this is true iff `0 <= m_index < bitset_type::s_bits_per_block`.

Could we test it?

Reply:
we can make assertions when we modify it